src/ts/background/workers/authorize.ts

```diff
@@ -0,0 +1,118 @@
+import {WorkerError, WorkerErrorKind} from "../../common/workers/types";
+import {BackgroundEvent} from "../../common/backgroundEvent";
+import {dispatchEvent} from "../util/dispatchEvent";
+import * as browser from "webextension-polyfill";
+import storage from "../../common/adapters/storage";
+
+type AuthorizeParameters = boolean;
+type AuthorizeResponse = string;
+
+type TokenRecord = {token: string; expiresAt: number};
+
+const isTokenRecord = (record: unknown): record is TokenRecord =>
+	record && typeof record === "object" && "token" in record && "expiresAt" in record;
+
+const parseAccessToken = (responseUrl: string): {token: string; expiresIn: number} => {
+	const {hash} = new URL(responseUrl);
+	const paramsString = hash.slice(1); // remove the "#" from the beginning
+	const params = new URLSearchParams(paramsString);
+	return {token: params.get("access_token"), expiresIn: Number(params.get("expires_in"))};
+};
+
+const getRedirectURL = (): string => {
+	const baseRedirectUrl = browser.identity.getRedirectURL();
+	if (baseRedirectUrl.includes("chromium")) {
+		return baseRedirectUrl;
+	} else {
+		const redirectSubdomain = baseRedirectUrl.match(/https:\/\/([^.]*)/)[1];
+		return `http://127.0.0.1/mozoauth2/${redirectSubdomain}`;
+	}
+};
+
+const getAuthUrl = () => {
+	const manifest = browser.runtime.getManifest() as any; // TODO there must be a way to fix this
+	const url = new URL("https://accounts.google.com/o/oauth2/auth");
+	const params = new URLSearchParams({
+		client_id: manifest.oauth2.client_id,
+		response_type: "token",
+		redirect_uri: getRedirectURL(),
+		scope: manifest.oauth2.scopes.join(" "),
+	});
+	url.search = params.toString();
+	return url.toString();
+};
+
+const setCachedToken = async (token: string, expiresIn: number): Promise<void> => {
+	const millisecondsRemaining = expiresIn * 1000;
+	const now = Date.now();
+	const fiveMinuteBuffer = 1000 * 60 * 5;
+	const expiresAt = now + millisecondsRemaining - fiveMinuteBuffer;
+	await storage.set("token", {token, expiresAt}); // TODO use satisfies
+};
+
+const getCachedToken = async (): Promise<string | undefined> => {
+	const record = await storage.get("token");
+	if (isTokenRecord(record)) {
+		const {token, expiresAt} = record;
+		if (Date.now() < expiresAt) {
+			return token;
+		} else {
+			await clearCachedToken();
```

Comment:
oh, so now that we're doing this, we have to invalidate the token ourselves?

does that mean that when it expires, we're effectively logging people out (we're not refreshing it for them)?

Reply:
Yeah it's the same as before, but we need to do it instead of Chrome handling it for us.